File: ontrack-extension-git/src/main/java/net/nemerosa/ontrack/extension/git/property/GitProjectConfigurationPropertyType.java
package net.nemerosa.ontrack.extension.git.property;

import com.fasterxml.jackson.databind.JsonNode;
import net.nemerosa.ontrack.common.MapBuilder;
import net.nemerosa.ontrack.extension.git.GitExtensionFeature;
import net.nemerosa.ontrack.extension.git.model.BasicGitConfiguration;
import net.nemerosa.ontrack.extension.git.service.GitConfigurationService;
import net.nemerosa.ontrack.model.form.Form;
import net.nemerosa.ontrack.model.form.Selection;
import net.nemerosa.ontrack.model.security.ProjectConfig;
import net.nemerosa.ontrack.model.security.SecurityService;
import net.nemerosa.ontrack.model.structure.ProjectEntity;
import net.nemerosa.ontrack.model.structure.ProjectEntityType;
import net.nemerosa.ontrack.model.support.ConfigurationPropertyType;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.EnumSet;
import java.util.Set;
import java.util.function.Function;

@Component
public class GitProjectConfigurationPropertyType extends AbstractGitProjectConfigurationPropertyType<GitProjectConfigurationProperty>
        implements ConfigurationPropertyType<BasicGitConfiguration, GitProjectConfigurationProperty> {

    private final GitConfigurationService configurationService;

    @Autowired
    public GitProjectConfigurationPropertyType(GitExtensionFeature extensionFeature, GitConfigurationService configurationService) {
        super(extensionFeature);
        this.configurationService = configurationService;
    }

    @Override
    public String getName() {
        return "Git configuration";
    }

    @Override
    public String getDescription() {
        return "Associates the project with a Git repository";
    }

    @Override
    public Set<ProjectEntityType> getSupportedEntityTypes() {
        return EnumSet.of(ProjectEntityType.PROJECT);
    }

    @Override
    public boolean canEdit(ProjectEntity entity, SecurityService securityService) {
        return securityService.isProjectFunctionGranted(entity.projectId(), ProjectConfig.class);
    }

    @Override
    public boolean canView(ProjectEntity entity, SecurityService securityService) {
        return true;
    }

    @Override
    public Form getEditionForm(ProjectEntity entity, GitProjectConfigurationProperty value) {
        return Form.create()
                .with(
                        Selection.of("configuration")
                                .label("Configuration")
                                .help("Git configuration to use to access the repository")
                                .items(configurationService.getConfigurationDescriptors())
                                .value(value != null ? value.getConfiguration().getName() : null)
                );
    }

    @Override
    public GitProjectConfigurationProperty fromClient(JsonNode node) {
        return fromStorage(node);
    }

    @Override
    public GitProjectConfigurationProperty fromStorage(JsonNode node) {
        String configurationName = node.path("configuration").asText();
        // Looks the configuration up
        BasicGitConfiguration configuration = configurationService.getConfiguration(configurationName);
        // OK
        return new GitProjectConfigurationProperty(
                configuration
        );
    }

    @Override
    public JsonNode forStorage(GitProjectConfigurationProperty value) {
        return format(
                MapBuilder.params()
                        .with("configuration", value.getConfiguration().getName())
                        .get()
        );
    }

    @Override
    public GitProjectConfigurationProperty replaceValue(GitProjectConfigurationProperty value, Function<String, String> replacementFunction) {
        return new GitProjectConfigurationProperty(
                configurationService.replaceConfiguration(value.getConfiguration(), replacementFunction)
        );
    }

}

File: ontrack-extension-git/src/main/java/net/nemerosa/ontrack/extension/git/property/GitProjectConfigurationPropertyType.kt
package net.nemerosa.ontrack.extension.git.property

import com.fasterxml.jackson.databind.JsonNode
import net.nemerosa.ontrack.extension.git.GitExtensionFeature
import net.nemerosa.ontrack.extension.git.model.BasicGitConfiguration
import net.nemerosa.ontrack.extension.git.service.GitConfigurationService
import net.nemerosa.ontrack.json.asJson
import net.nemerosa.ontrack.model.security.ProjectConfig
import net.nemerosa.ontrack.model.security.SecurityService
import net.nemerosa.ontrack.model.structure.ProjectEntity
import net.nemerosa.ontrack.model.structure.ProjectEntityType
import net.nemerosa.ontrack.model.support.ConfigurationPropertyType
import org.springframework.stereotype.Component
import java.util.*


@Component
@Deprecated("Will be removed in V5. Pure Git configuration won't be supported any longer.")
class GitProjectConfigurationPropertyType(
    extensionFeature: GitExtensionFeature,
    private val configurationService: GitConfigurationService
) : AbstractGitProjectConfigurationPropertyType<GitProjectConfigurationProperty>(extensionFeature),
    ConfigurationPropertyType<BasicGitConfiguration, GitProjectConfigurationProperty> {

    override val name: String = "Git configuration"

    override val description: String = "Associates the project with a Git repository"

    override val supportedEntityTypes: Set<ProjectEntityType> = EnumSet.of(ProjectEntityType.PROJECT)

    override fun canEdit(entity: ProjectEntity, securityService: SecurityService): Boolean {
        return securityService.isProjectFunctionGranted(entity.projectId(), ProjectConfig::class.java)
    }

    override fun canView(entity: ProjectEntity, securityService: SecurityService): Boolean {
        return true
    }

    override fun fromClient(node: JsonNode): GitProjectConfigurationProperty {
        return fromStorage(node)
    }

    override fun fromStorage(node: JsonNode): GitProjectConfigurationProperty {
        val configurationName = node.path("configuration").asText()
        // Looks the configuration up
        val configuration = configurationService.getConfiguration(configurationName)
        // OK
        return GitProjectConfigurationProperty(
            configuration
        )
    }

    override fun forStorage(value: GitProjectConfigurationProperty): JsonNode =
        mapOf("configuration" to value.configuration.name)
            .asJson()

    override fun replaceValue(
        value: GitProjectConfigurationProperty,
        replacementFunction: (String) -> String
    ): GitProjectConfigurationProperty {
        return GitProjectConfigurationProperty(
            value.configuration
        )
    }
}
